src/pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -486,10 +494,10 @@ class _DataLoaderSource:
             that returns the desired dataloader(s).
     """
 
-    instance: Optional[Union[TRAIN_DATALOADERS, EVAL_DATALOADERS, "pl.LightningModule", "pl.LightningDataModule"]]
+    instance: INSTANCE
     name: str
 
-    def dataloader(self) -> Union[TRAIN_DATALOADERS, EVAL_DATALOADERS]:
+    def dataloader(self) -> Union[INSTANCE, TRAIN_DATALOADERS, EVAL_DATALOADERS]:
```

Comment:
The original types were correct, this method cannot return `LightningModule` or `LightningDataModule`. Either resolve it in the method (which I'm not really sure how feasible it is, since that method does a bit of magic, that `mypy` might not like) or add proper `# type: ignore`s, but don't extend the result type of this method

Reply:
I opted to ignore after attempting to use self.is_defined to block mypy from evaluating return-value as a LightningModule.